File: crates/klio-cfa/src/builder.rs
//! Low-level CFG construction primitives. The AST → CFG lowering
//! pass (`lower.rs`) drives this builder; tests construct CFGs
//! directly through it. The builder is intentionally dumb: it
//! allocates blocks, registers, and edges; it does not know about
//! the source language.

use crate::ir::{BasicBlock, BlockId, Cfg, Edge, EdgeKind, LabelId, LoopId, Node, Reg, Terminator};
use klio_span::Span;

pub struct CfgBuilder {
    blocks: Vec<BasicBlock>,
    next_reg: u32,
    next_loop: u32,
    next_label: u32,
}

impl CfgBuilder {
    #[must_use]
    pub fn new() -> Self {
        Self {
            blocks: Vec::new(),
            next_reg: 0,
            next_loop: 0,
            next_label: 0,
        }
    }

    /// # Panics
    /// Panics if the block count exceeds `u32::MAX`.
    pub fn new_block(&mut self) -> BlockId {
        let id = BlockId(u32::try_from(self.blocks.len()).expect("block id overflow"));
        self.blocks.push(BasicBlock {
            id,
            nodes: Vec::new(),
            term: Terminator::Unreachable,
            preds: Vec::new(),
            succs: Vec::new(),
        });
        id
    }

    /// # Panics
    /// Panics if the register count exceeds `u32::MAX`.
    pub fn new_reg(&mut self) -> Reg {
        let r = Reg(self.next_reg);
        self.next_reg = self.next_reg.checked_add(1).expect("reg id overflow");
        r
    }

    /// # Panics
    /// Panics if the loop count exceeds `u32::MAX`.
    pub fn new_loop(&mut self) -> LoopId {
        let id = LoopId(self.next_loop);
        self.next_loop = self.next_loop.checked_add(1).expect("loop id overflow");
        id
    }

    /// # Panics
    /// Panics if the label count exceeds `u32::MAX`.
    pub fn new_label(&mut self) -> LabelId {
        let id = LabelId(self.next_label);
        self.next_label = self.next_label.checked_add(1).expect("label id overflow");
        id
    }

    /// Append a node to the given block. Order-sensitive.
    pub fn push(&mut self, blk: BlockId, node: Node) {
        self.blocks[blk.0 as usize].nodes.push(node);
    }

    /// Number of nodes already in `blk`. Useful for callers that
    /// want to remember the position a node is about to be pushed
    /// into — the returned value is the insertion index.
    #[must_use]
    pub fn current_node_count(&self, blk: BlockId) -> Option<usize> {
        self.blocks.get(blk.0 as usize).map(|b| b.nodes.len())
    }

    /// Set the terminator for a block and wire up the edges to the
    /// referenced successors. Replaces any prior terminator on
    /// `blk` and any preds/succs implied by it.
    pub fn set_terminator(&mut self, blk: BlockId, term: Terminator) {
        self.unwire_succs(blk);
        let succs: Vec<Edge> = match &term {
            Terminator::Goto(t) => vec![Edge {
                block: *t,
                kind: EdgeKind::Normal,
            }],
            Terminator::Branch {
                then_blk, else_blk, ..
            } => vec![
                Edge {
                    block: *then_blk,
                    kind: EdgeKind::True,
                },
                Edge {
                    block: *else_blk,
                    kind: EdgeKind::False,
                },
            ],
            Terminator::Switch { arms, default, .. } => {
                let mut v: Vec<Edge> = arms
                    .iter()
                    .map(|a| Edge {
                        block: a.target,
                        kind: EdgeKind::Normal,
                    })
                    .collect();
                v.push(Edge {
                    block: *default,
                    kind: EdgeKind::Normal,
                });
                v
            }
            Terminator::Throw(_) | Terminator::Return(_) | Terminator::Unreachable => Vec::new(),
        };
        self.blocks[blk.0 as usize].term = term;
        for edge in &succs {
            self.blocks[edge.block.0 as usize].preds.push(Edge {
                block: blk,
                kind: edge.kind.clone(),
            });
        }
        self.blocks[blk.0 as usize].succs = succs;
    }

    /// Add an additional out-edge of an arbitrary kind from `from` to
    /// `to`. Used for exception edges and finally entry/exit, which
    /// are not implied by the terminator shape.
    pub fn add_edge(&mut self, from: BlockId, to: BlockId, kind: EdgeKind) {
        self.blocks[from.0 as usize].succs.push(Edge {
            block: to,
            kind: kind.clone(),
        });
        self.blocks[to.0 as usize]
            .preds
            .push(Edge { block: from, kind });
    }

    /// Detach any preds that point to `blk`'s prior terminator-implied
    /// successors so we can replace them cleanly.
    fn unwire_succs(&mut self, blk: BlockId) {
        let prior: Vec<Edge> = std::mem::take(&mut self.blocks[blk.0 as usize].succs);
        for edge in prior {
            let dst = &mut self.blocks[edge.block.0 as usize].preds;
            if let Some(idx) = dst
                .iter()
                .position(|e| e.block == blk && e.kind == edge.kind)
            {
                dst.swap_remove(idx);
            }
        }
    }

    #[must_use]
    pub fn finish(self, entry: BlockId, exits: Vec<BlockId>, source: Span) -> Cfg {
        Cfg {
            blocks: self.blocks,
            entry,
            exits,
            source,
            next_reg: self.next_reg,
        }
    }
}

impl Default for CfgBuilder {
    fn default() -> Self {
        Self::new()
    }
}

File: src/cfa/builder.zig
//! Low-level CFG construction primitives, driven by the lowering pass and by
//! tests that build CFGs directly. Allocates blocks, registers and edges and
//! knows nothing of the source language.

const std = @import("std");
const span = @import("span");
const ir = @import("ir.zig");

const Allocator = std.mem.Allocator;
const Span = span.Span;
const BasicBlock = ir.BasicBlock;
const BlockId = ir.BlockId;
const Cfg = ir.Cfg;
const Edge = ir.Edge;
const EdgeKind = ir.EdgeKind;
const LabelId = ir.LabelId;
const LoopId = ir.LoopId;
const Node = ir.Node;
const Reg = ir.Reg;
const Terminator = ir.Terminator;

pub const CfgBuilder = struct {
    blocks: std.ArrayList(BasicBlock) = .empty,
    next_reg: u32 = 0,
    next_loop: u32 = 0,
    next_label: u32 = 0,

    pub fn init() CfgBuilder {
        return .{};
    }

    pub fn newBlock(self: *CfgBuilder, allocator: Allocator) Allocator.Error!BlockId {
        const id = BlockId.from(@intCast(self.blocks.items.len));
        try self.blocks.append(allocator, .{ .id = id });
        return id;
    }

    pub fn newReg(self: *CfgBuilder) Reg {
        const r = Reg.from(self.next_reg);
        self.next_reg += 1;
        return r;
    }

    pub fn newLoop(self: *CfgBuilder) LoopId {
        const id = LoopId.from(self.next_loop);
        self.next_loop += 1;
        return id;
    }

    pub fn newLabel(self: *CfgBuilder) LabelId {
        const id = LabelId.from(self.next_label);
        self.next_label += 1;
        return id;
    }

    pub fn push(self: *CfgBuilder, allocator: Allocator, blk: BlockId, node: Node) Allocator.Error!void {
        try self.blocks.items[blk.int()].nodes.append(allocator, node);
    }

    /// The index the next pushed node lands at.
    pub fn currentNodeCount(self: *const CfgBuilder, blk: BlockId) ?usize {
        if (blk.int() >= self.blocks.items.len) return null;
        return self.blocks.items[blk.int()].nodes.items.len;
    }

    /// Wires the edges to the successors, replacing any prior terminator and
    /// the preds/succs it implied.
    pub fn setTerminator(self: *CfgBuilder, allocator: Allocator, blk: BlockId, term: Terminator) Allocator.Error!void {
        self.unwireSuccs(blk);
        var succs: std.ArrayList(Edge) = .empty;
        errdefer succs.deinit(allocator);
        switch (term) {
            .Goto => |t| try succs.append(allocator, .{ .block = t, .kind = .Normal }),
            .Branch => |b| {
                try succs.append(allocator, .{ .block = b.then_blk, .kind = .True });
                try succs.append(allocator, .{ .block = b.else_blk, .kind = .False });
            },
            .Switch => |sw| {
                for (sw.arms) |a| {
                    try succs.append(allocator, .{ .block = a.target, .kind = .Normal });
                }
                try succs.append(allocator, .{ .block = sw.default, .kind = .Normal });
            },
            .Throw, .Return, .Unreachable => {},
        }
        self.blocks.items[blk.int()].term = term;
        for (succs.items) |edge| {
            try self.blocks.items[edge.block.int()].preds.append(allocator, .{
                .block = blk,
                .kind = edge.kind,
            });
        }
        self.blocks.items[blk.int()].succs.deinit(allocator);
        self.blocks.items[blk.int()].succs = succs;
    }

    /// For the exception and finally edges a terminator shape does not imply.
    pub fn addEdge(self: *CfgBuilder, allocator: Allocator, from: BlockId, to: BlockId, kind: EdgeKind) Allocator.Error!void {
        try self.blocks.items[from.int()].succs.append(allocator, .{ .block = to, .kind = kind });
        try self.blocks.items[to.int()].preds.append(allocator, .{ .block = from, .kind = kind });
    }

    /// Detach the preds pointing at `blk`'s prior successors so they can be replaced cleanly.
    fn unwireSuccs(self: *CfgBuilder, blk: BlockId) void {
        const prior = &self.blocks.items[blk.int()].succs;
        for (prior.items) |edge| {
            const dst = &self.blocks.items[edge.block.int()].preds;
            var idx: usize = 0;
            while (idx < dst.items.len) : (idx += 1) {
                const e = dst.items[idx];
                if (e.block == blk and e.kind.eql(edge.kind)) {
                    _ = dst.swapRemove(idx);
                    break;
                }
            }
        }
        prior.clearRetainingCapacity();
    }

    pub fn finish(
        self: *CfgBuilder,
        allocator: Allocator,
        entry: BlockId,
        exits: std.ArrayList(BlockId),
        source: Span,
    ) Cfg {
        _ = allocator;
        return .{
            .blocks = self.blocks,
            .entry = entry,
            .exits = exits,
            .source = source,
            .next_reg = self.next_reg,
        };
    }
};

test "builder allocates blocks and registers" {
    const a = std.testing.allocator;
    var b = CfgBuilder.init();
    defer {
        for (b.blocks.items) |*blk| {
            blk.nodes.deinit(a);
            blk.preds.deinit(a);
            blk.succs.deinit(a);
        }
        b.blocks.deinit(a);
    }
    const b0 = try b.newBlock(a);
    const b1 = try b.newBlock(a);
    try std.testing.expectEqual(@as(u32, 0), b0.int());
    try std.testing.expectEqual(@as(u32, 1), b1.int());
    const r0 = b.newReg();
    const r1 = b.newReg();
    try std.testing.expectEqual(@as(u32, 0), r0.int());
    try std.testing.expectEqual(@as(u32, 1), r1.int());
    try b.setTerminator(a, b0, .{ .Goto = b1 });
    try std.testing.expectEqual(@as(usize, 1), b.blocks.items[b0.int()].succs.items.len);
    try std.testing.expectEqual(@as(usize, 1), b.blocks.items[b1.int()].preds.items.len);
}
